### PR TITLE
Fix ORTTrainer failure on gpt2 fp16 training

### LINKED ISSUE #11279: Type Error when training Hugging Face Transformers GPT2 with fp16 enabled
**Describe the bug**
During training of a huggingface transformers (version 4.18.0) GPT2 with onnxruntime training (model is wrapped in ORTModule) and fp16 mixed precision training enabled, I encounter a type error.

```
RuntimeError: /onnxruntime_src/orttraining/orttraining/python/orttraining_pybind_state.cc:752 
onnxruntime::python::addObjectMethodsForTraining(pybind11::module&, 
onnxruntime::python::ExecutionProviderRegistrationFn)::<lambda(onnxruntime::training::OrtModuleGraphBuilder*, 
const pybind11::bytes&, const onnxruntime::training::OrtModuleGraphBuilderConfiguration&)> 
[ONNXRuntimeError] : 1 : FAIL : Type Error: Type parameter (T) of Optype (Where) bound to different types 
(tensor(float) and tensor(float16) in node (Where_199). 
```

I do not encounter this issue when using transformers version 4.16.0.

This bug is similar to the one reported in issue https://github.com/pytorch/ort/issues/104.  That issue has been closed but it appears to me that the mixed precision training issue hasn't actually been resolved. The conversation on the issue changed to some other errors encountered while moving to `onnxruntime-training 1.11.0`. It was closed after the new error was resolved. However, I still encounter the same mixed precision training issue when using the recommended package versions.

**Urgency**
We are considering using huggingface optimum ORTTrainer to integrate ORT training into huggingface transformers. 
Migration to optimum is blocked by this issue.

**System information**
- torch 1.11.0 
- transformers 4.18.0 
- deepspeed 0.6.1
- torch-ort 1.11.0
- onnx 1.10.2
- onnxruntime-training 1.11.0+cu113
- GPU: v100 (16GiB)

**To Reproduce**
The code and instructions for reproduction can be found in this [Git repo directory ](https://github.com/jambayk/transformers-ort-failures/tree/main/gpt2). 

Please contact me (jambaykinley@microsoft.com) for access to the repo.

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

Fixes [#11279 of onnxruntime](https://github.com/microsoft/onnxruntime/issues/11279)

## Context

Optimum users reported that the mixed-precision training on gpt2 with `optimum.onnxruntime.ORTTrainer` is broken since transformers>4.16.0. After investigation, the break comes from the [removal of `float()`](https://github.com/huggingface/transformers/pull/14321/files#r912791693) in gpt2 modeling from PR #14321.

__Reproduction__

Run optimum onnxruntime training example [run_glue.py](https://github.com/huggingface/optimum/blob/main/examples/onnxruntime/training/text-classification/run_glue.py) with:
```bash
python run_glue.py \
    --model_name_or_path gpt2 \
    --task_name sst2 \
    --do_train \
    --do_eval \
    --fp16 \
    --output_dir /tmp/ort-gpt2-sst2/
```
 
__Error Message__
```
RuntimeError: /onnxruntime_src/orttraining/orttraining/python/orttraining_pybind_state.cc:752 
onnxruntime::python::addObjectMethodsForTraining(pybind11::module&, 
onnxruntime::python::ExecutionProviderRegistrationFn)::<lambda(onnxruntime::training::OrtModuleGraphBuilder*, const 
pybind11::bytes&, const onnxruntime::training::OrtModuleGraphBuilderConfiguration&)> [ONNXRuntimeError] : 1 : FAIL : 
Type Error: Type parameter (T) of Optype (Where) bound to different types (tensor(float) and tensor(float16) in node 
(Where_201).
```
As mentioned in the error message, the forward with onnxruntime InferenceSession will fail on a node __Where__ in the graph, which corresponds to the [Where op](https://github.com/michaelbenayoun/transformers/blob/9ef5813d73b2af238e8820e117dd32415ce4c173/src/transformers/models/gpt2/modeling_gpt2.py#L206) in gpt2 modeling.

And the problem comes from the fact that after removing `float()`, during fp16 training, the inputs of Where have different dtype (one in fp32 and one in fp16), which violates the definition in ONNX and leads to the failure.

## Who can review?

@michaelbenayoun @patrickvonplaten, @LysandreJik

## Fix

Ensure `attn_weights` and `value` has the same type in exported ONNX IR.

### CODE REVIEW

**Review by michaelbenayoun — APPROVED** [review]
LGTM (once all the tests pass)

**Comment by ydshieh** [comment]
Hello @JingyaHuang By looking this 2 lines

https://github.com/huggingface/transformers/blob/04ffba9a9a4d452b328eb7e57f7c87f2ed78bd10/src/transformers/models/decision_transformer/modeling_decision_transformer.py#L196-L197

`mask_value` is already of type `attn_weights.dtype`. Does the issue only occur when we use ONNX? (i.e. if we run the model in Python with FP16, does it work?). This issue seems strange. Do you happen to know which argument gets fp32 and which one gets fp16?

**Comment by JingyaHuang** [comment]
Hi @ydshieh, yes this issue only occurs with ONNX. When exporting the ONNX IR `mask_value` is exported as a constant initializer(min of fp32) with dtype float32. Thus during the mixed-precision training with onnxruntime, the `attn_weights` will be in dtype fp16 and `mask_value` as a constant always fp32 -> two inputs with different dtype -> training failed. 
Here is the ONNX IR which illustrates what happened with Where op:
image

**Comment by ydshieh** [comment]
@JingyaHuang Thank you!

File: src/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -189,7 +189,9 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
         if self.scale_attn_weights:
-            attn_weights = attn_weights / (value.size(-1) ** 0.5)
+            attn_weights = attn_weights / torch.tensor(value.size(-1) ** 0.5, dtype=attn_weights.dtype).to(
```
Comment (ydshieh) [inline src/transformers/models/gpt2/modeling_gpt2.py:192]: So this will make `attn_weights` becomes FP32 in ONNX at the `where` OP?

Comment (ydshieh) [reply src/transformers/models/gpt2/modeling_gpt2.py:192]: I feel like the situation is more like:

**without the fix:**
`attn_weights` will be in dtype fp32 (at where op) and `mask_value` as a constant always fp16

**with the fix:**
both in fp16

But please correct me as I know very little here (just by intuition)

**Comment by JingyaHuang** [comment]
Gently pinging @patrickvonplaten and @LysandreJik for a review.

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py [hunk 2]
```diff
@@ -178,7 +178,9 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         attn_weights = torch.matmul(query, key.transpose(-1, -2))
 
         if self.scale_attn_weights:
-            attn_weights = attn_weights / (value.size(-1) ** 0.5)
+            attn_weights = attn_weights / torch.tensor(value.size(-1) ** 0.5, dtype=attn_weights.dtype).to(
+                attn_weights.device
+            )
```
Comment (LysandreJik) [inline src/transformers/models/decision_transformer/modeling_decision_transformer.py:183]: Nit, how about something like this @JingyaHuang ?

Other than that, and the same below, happy to merge it

```suggestion
            attn_weights = attn_weights / torch.tensor(value.size(-1) ** 0.5, dtype=attn_weights.dtype, device=attn_weights.device)
```

**Review by LysandreJik — APPROVED** [review]

File: src/transformers/models/decision_transformer/modeling_decision_transformer.py (hunk 2 above)
Comment (JingyaHuang) [reply src/transformers/models/decision_transformer/modeling_decision_transformer.py]: Fixed for both, thanks for the suggestion!

**Review by LysandreJik — APPROVED** [review]
Thanks, merging!

**Comment by TXacs** [comment]
I got the same error. But I use the torch.fx and amp to train the GPT2 model. I fix this error with the method is add `attn_weights.to(attn_weights.dtype)` in `torch.where`. I don't know why this way can fix it, but it does.

**Comment by ydshieh** [comment]
Hi @TXacs , which version you tried? Could you try to install the latest version on `main`: `pip install git+https://github.com/huggingface/accelerate`, and see if you still have the issue (without your fix)

**Pull request merged**